### ISSUE TITLE
Uploading folder with a lot of sub-folders causes Chrome to hang/crash

### ISSUE THREAD

**Issue by nickdatum** [23/resumable.js#310]
Take a look at this jsFiddle and then try dragging a folder with a large number of files in it and it causes the browser to hang (and usually crash)

https://jsfiddle.net/ad3q077k/4/

The folder I was testing with comprised of 6,286 Files across 1,150 Folders

It didn't seem to have the issue when uploading a folder with just 4-sub-folders and 9000+ files (some of which were really large).

So I suspect it's the folder structure causing the crash...perhaps the fact it's recursively calling readEntries causing the hangs/crashes.

I realise sending you the folder I'm testing with isn't possible, but you can replicate fairly easily by simply uploading a system folder or something (e.g. in windows the "windows" or "program files" folders).

I thought this may be a browser bug, but I noticed it works fine in Google Drive with no crashes/hangs.

**Comment by steffentchr** [comment]
Hi @nickdatum -- I'd accept a pull request to improve this performance, but since Resumable.js at its core is designed to handle large files resiliently, the scale of file handling seems somewhat out of scope. Obviously, if you're looking at it and find good solutions, please share.

**Comment by nickdatum** [comment]
I have an idea with this... I think it's because the files array is getting too large and filling up the memory. I also suspect there is some duplicates being put into the memory, but I'll have to debug further first to figure this all out.

I may have a play later in the week and potentially try to defer some of the processing so it doesn't iterate through directories all in one go before starting the upload.

Do you think this could work?

**Comment by steffentchr** [comment]
It's very likely that -- combined with the fact that the library does a lot of operations on "all files at the same time". Adding 10K files will create as many instances of ResumableFile immediately, and all calls to methods like `.progress()` will loop through all and continuously poll the progress of each items. Which is going to be a pain.

The problem is certainly solvable; I'm worried that it's a pretty sizeable rewrite of lib though. On hacky solution would be to manage drag/drop/files array outside of Resumable with a queue that would feed files to Resumable in drips rather than all at once.

**Comment by nickdatum** [comment]
I'm going to spend the day today doing some memory profiling and seeing if I can make any sort of performance improvements...could end up being a waste of a day or a really productive one, I'll keep you updated!

**Comment by nickdatum** [comment]
@steffentchr I've rewrote the loadFiles function and it works a lot more efficiently now. I also removed the incorrect paths patch I'd applied before as this also fixed this as part of the rewrite (issue #304)

Please let me know what you think, I think you'll be pretty impressed with the performance increase! The main benefit is it doesn't cause the browser to completely hang (and sometimes crash) so you can still use whatever web application you're integrating resumable into.

https://github.com/nickdatum/resumable.js/blob/performance-relative/resumable.js

I've also added a new "beforeAdd" event, as quite often users will want to show a message or something to say files are currently being loaded in for processing (or other possible actions).

**Comment by steffentchr** [comment]
That look great -- very solid improvements from the look of the code, and it seems to be doing all the optimisations in a transparent manner that isn't even breaking backwards-compatibility. Any outstanding things or do you want to merge this in to the core liv?

**Comment by nickdatum** [comment]
Thanks, glad it's working for you now.

There was one thing breaking with backwards compatibility (missing forward slash at beginning of relative path) but I've sorted this now in my latest push.

Happy for you to merge in now and glad to help out. I made use of the project for my app, so it felt good to contribute something back :)

**Comment by steffentchr** [comment]
Awesome -- do you want to submit as a pull request?

**Comment by nickdatum** [comment]
Ok, just submitted the pull request

**Comment by steffentchr** [comment]
This one, right? https://github.com/23/resumable.js/pull/313

**Comment by nickdatum** [comment]
That's the one! Thanks
